[visionOS] View Spatial Photos on the Web with zero author adoption

yoelhawa opened the pull request:
#### 6a05886107b11dbc1c96cee394a1d7f986489e9f
<pre>
[visionOS] View Spatial Photos on the Web with zero author adoption
<a href="https://bugs.webkit.org/show_bug.cgi?id=277447">https://bugs.webkit.org/show_bug.cgi?id=277447</a>
&lt;<a href="https://rdar.apple.com/130545126">rdar://130545126</a>&gt;

Reviewed by Aditya Keerthi.

Add a new context menu item to spatial photos on visionOS to view them
spatially.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Introduce new flag for feature.

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::isSpatial const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::isSpatial const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::isSpatial const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::isSpatial const):
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.mm:
(-[_WKActivatedElementInfo _initWithInteractionInformationAtPosition:isUsingAlternateURLForImage:userInfo:]):
(-[_WKActivatedElementInfo isSpatialImage]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::imagePositionInformation):
Plumb spatial photo detection from image load to hit-test to activated
element.

* Source/WebKit/Shared/ios/GestureTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(+[_WKElementAction _elementActionWithType:customTitle:assistant:disabled:]):
(+[_WKElementAction imageForElementActionType:]):
(elementActionTypeToUIActionIdentifier):
(uiActionIdentifierToElementActionType):
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
(-[WKActionSheetAssistant handleElementActionWithType:element:needsInteraction:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::performActionOnElement):
* Source/WebCore/en.lproj/Localizable.strings:
Add context menu item to spatial photos to view the photo spatially.

Canonical link: <a href="https://commits.webkit.org/282016@main">https://commits.webkit.org/282016@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f7f0c881109e09f85e31ee8f19b2cfe1719cfc76

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/61756 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/41110 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/14348 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/65735 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/12301 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/63875 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/48796 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/12572 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/49807 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wincairo-tests~~](https://ews-build.webkit.org/#/builders/60/builds/8541 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/64825 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/38187 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/53494 "Passed tests") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/30639 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/34844 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/10724 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/11232 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/56649 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/63/builds/11027 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/67463 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/5699 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/62/builds/10785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/57186 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/5724 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/53441 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/57419 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/4689 "Passed tests") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/9307 "Built successfully and passed tests") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/36910 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/37994 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/39090 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
| | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/37739 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

etiennesegonzac commented on Source/WebKit/Shared/ios/GestureTypes.h:
```diff
@@ -60,7 +60,8 @@ enum class SheetAction : uint8_t {
     PauseAnimation,
-    PlayAnimation
+    PlayAnimation,
+    ViewSpatial
```
should it be gated by a `if ENABLE(SPATIAL_IMAGE_DETECTION)`?

yoelhawa replied:
Done, but not marking resolved in case anyone weighs in on why there are no other compilation flags in this file (it's possible I'm just first!)

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.h:
```diff
@@ -48,6 +48,7 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
 @property (nonatomic, readonly) CGRect boundingRect;
 @property (nonatomic, readonly) NSString *ID WK_API_AVAILABLE(macos(10.12), ios(10.0));
 @property (nonatomic, readonly) BOOL isAnimatedImage WK_API_AVAILABLE(macos(10.15), ios(13.0));
+@property (nonatomic, readonly) BOOL isSpatialImage WK_API_AVAILABLE(visionos(WK_XROS_TBA));
```
should be gated too to match the implementation

yoelhawa replied:
Note that I get `error: function-like macro 'ENABLE' is not defined` so I'd have to use the standard `ENABLE_ SPATIAL_IMAGE_DETECTION` flag here. I know in the past I have broken builds with using an `ENABLE_...` flag that way so I'd like to make sure that won't happen again

yoelhawa replied:
Aaand looks like this did in fact break the build (errors about that flag not being defined) so I’m removing it again

etiennesegonzac replied:
Hum... can you put it in `_WKActivatedElementInfoInternal.h`, with the flag?

pxlcoder replied:
You can't use our macros in headers that are part of the SDK (both public and private).

But you could use `TARGET_OS_VISION`.

yoelhawa replied:
Added `if defined(TARGET_OS_VISION) && TARGET_OS_VISION` here

etiennesegonzac commented on Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
```diff
@@ -160,6 +161,14 @@ + (instancetype)_elementActionWithType:(_WKElementActionType)type customTitle:(N
             [assistant handleElementActionWithType:type element:actionInfo needsInteraction:YES];
         };
         break;
+    case _WKElementActionTypeViewSpatial:
```
we have both styles in this file, but since this is brand new I'd err on the side of having everything behind the flag (including the case)

yoelhawa replied:
I can in this particular instance, but in the other two switches in this file where `_WKElementActionTypeViewSpatial` appears I have to leave the case outside the flag or EWS yells at me for not handling all enumeration cases.

etiennesegonzac reviewed (commented):
a couple of comments while we wait for a reviewer :)

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.h:
```diff
@@ -56,6 +56,7 @@ typedef NS_ENUM(NSInteger, _WKElementActionType) {
     _WKElementActionTypeImageExtraction WK_API_AVAILABLE(ios(15.0)),
     _WKElementActionTypeRevealImage WK_API_AVAILABLE(ios(15.0)),
     _WKElementActionTypeCopyCroppedImage WK_API_AVAILABLE(ios(16.0)),
+    _WKElementActionTypeViewSpatial WK_API_AVAILABLE(visionos(WK_XROS_TBA)),
```
I think this can also be behind `TARGET_OS_VISION`?

pxlcoder commented on Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
```diff
@@ -248,6 +248,15 @@ bool BitmapImageDescriptor::shouldUseQuickLookForFullscreen() const
+#if ENABLE(SPATIAL_IMAGE_DETECTION)
+bool BitmapImageDescriptor::isSpatial() const
+{
+    if (auto decoder = m_source.decoderIfExists())
```
```suggestion
    if (RefPtr decoder = m_source.decoderIfExists())
```

pxlcoder commented on Source/WTF/wtf/PlatformEnableCocoa.h:
```diff
@@ -1056,3 +1056,7 @@
+
+#if !defined(ENABLE_SPATIAL_IMAGE_DETECTION) && PLATFORM(VISION)
+#define ENABLE_SPATIAL_IMAGE_DETECTION 0
+#endif
```
You can actually just omit this entirely. `ENABLE()` will automatically evaluate to 0 for anything that isn't defined.

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
```diff
@@ -51,6 +51,7 @@
 static UIActionIdentifier const WKElementActionTypeOpenIdentifier = @"WKElementActionTypeOpen";
 static UIActionIdentifier const WKElementActionTypeCopyIdentifier = @"WKElementActionTypeCopy";
 static UIActionIdentifier const WKElementActionTypeSaveImageIdentifier = @"WKElementActionTypeSaveImage";
+static UIActionIdentifier const WKElementActionTypeViewSpatialIdentifier = @"WKElementActionTypeViewSpatial";
```
This should probably be guarded too?

pxlcoder commented on Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.h:
```diff
@@ -48,6 +48,9 @@ WK_CLASS_AVAILABLE(macos(10.10), ios(8.0))
 @property (nonatomic, readonly) CGRect boundingRect;
 @property (nonatomic, readonly) NSString *ID WK_API_AVAILABLE(macos(10.12), ios(10.0));
 @property (nonatomic, readonly) BOOL isAnimatedImage WK_API_AVAILABLE(macos(10.15), ios(13.0));
+#if defined(TARGET_OS_VISION) && TARGET_OS_VISION
```
I think this should have `__VISION_OS_VERSION_MIN_REQUIRED` to match?

Commit:
[visionOS] View Spatial Photos on the Web with zero author adoption
https://bugs.webkit.org/show_bug.cgi?id=277447
<rdar://130545126>

Reviewed by Aditya Keerthi.

Add a new context menu item to spatial photos on visionOS to view them
spatially.

* Source/WTF/wtf/PlatformEnableCocoa.h:
Introduce new flag for feature.

* Source/WebCore/platform/graphics/BitmapImage.h:
* Source/WebCore/platform/graphics/BitmapImageDescriptor.cpp:
(WebCore::BitmapImageDescriptor::isSpatial const):
* Source/WebCore/platform/graphics/BitmapImageDescriptor.h:
* Source/WebCore/platform/graphics/BitmapImageSource.h:
* Source/WebCore/platform/graphics/Image.h:
(WebCore::Image::isSpatial const):
* Source/WebCore/platform/graphics/ImageDecoder.h:
(WebCore::ImageDecoder::isSpatial const):
* Source/WebCore/platform/graphics/ImageSource.h:
(WebCore::ImageSource::isSpatial const):
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.h:
* Source/WebKit/Shared/ios/InteractionInformationAtPosition.serialization.in:
* Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKActivatedElementInfo.mm:
(-[_WKActivatedElementInfo _initWithInteractionInformationAtPosition:isUsingAlternateURLForImage:userInfo:]):
(-[_WKActivatedElementInfo isSpatialImage]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::imagePositionInformation):
Plumb spatial photo detection from image load to hit-test to activated
element.

* Source/WebKit/Shared/ios/GestureTypes.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKElementAction.mm:
(+[_WKElementAction _elementActionWithType:customTitle:assistant:disabled:]):
(+[_WKElementAction imageForElementActionType:]):
(elementActionTypeToUIActionIdentifier):
(uiActionIdentifierToElementActionType):
* Source/WebKit/UIProcess/ios/WKActionSheetAssistant.mm:
(-[WKActionSheetAssistant defaultActionsForImageSheet:]):
(-[WKActionSheetAssistant handleElementActionWithType:element:needsInteraction:]):
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::performActionOnElement):
* Source/WebCore/en.lproj/Localizable.strings:
Add context menu item to spatial photos to view the photo spatially.

Canonical link: https://commits.webkit.org/282016@main
